### ISSUE TITLE
Investigate Asynchronous Implications and Conversion

### ISSUE THREAD

**Issue by WithoutAnAccent** [A1tered/SimpleAntivirus#283]
<h1>(Issue pretty much concerns every module)</h1>
Main Concerns: Alerts, GUI, Every other module)

I have notable concerns about integration with UI now. 

After reading many asynchronous resources, integration with UI is a major concern!

Current implementation of Integrity Checking and File Hash Checking are not asynchronous, they could potentially halt the UI thread. Both modules can run functions in parallel for speed but they do not have Asynchronous functions.

Eg, running of Integrity Checking / File Hash Checking could freeze the program when running scans and what not. 

It is assumed that Integrity checking and File hash checking can be scanning simultaneously, another issue arises, Alerts and its reliance on events. Could utilizing Alerts on different threads cause issues (Implications unknown)?

Another notable concern on top of Asynchronous programming is that Events are synchronous and Alerts rely on events. Meaning that If I'm running Integrity Checking scan, the program will HALT to send an alert to Alerts. Meaning that if Alerts is going to do something big then it could halt the scan all together.

I am concerned of the stack call involved here, and want to highlight it and draw more research upon it.

Eg.
UI -> Integrity Checking/File Hash Scan -> Scan Files (Loop) -> (Event Raise) -> Alerts -> UI + Database Operations

                                                          (Loop stops to service the Event)

(Whilst it is not a major issue, it must be noted that Alerts should probably never call back into the Main System, eg, Alerts afterwards triggers another Scan or what not)

Major Concerns:
* Effect of scans on the UI thread (https://codingjourneyman.com/2015/10/05/async-await-and-the-ui-thread/)
* Events to Alerts (Synchronous based)
* Multiple events send to Alerts at once from different threads (Eg. Integrity Checking / File hash checking)

Maybe Considerations (Im no expert but these are considerations):
* Events to Alerts add to a **ConcurrentQueue** (thread safe queue). In an **outer** background loop, Alerts checks this queue for alerts. (Issues: No idea how to do a background task that checks for this, but it seems potentially safer)

By doing this, IntegrityChecking/File hash checking only have to worry about adding to Alerts to the queue, and don't have alerts do anything in its thread / stack.`
